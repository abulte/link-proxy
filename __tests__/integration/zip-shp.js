const nock = require('nock')

const mongo = require('../../lib/utils/mongo')
const {upsertLink, getLinkSummary} = require('../../lib/link')
const check = require('../../jobs/check')

const {shapefile} = require('../__helpers__/archive')
const {downloadsSnapshot} = require('../__helpers__/snapshots')

const NAME = 'test-link-proxy-zip-shp'

jest.mock('bull-manager', () => ({
  enqueue: jest.fn()
}))

beforeAll(async () => {
  process.env.MONGO_DB = NAME

  await mongo.connect()
  await mongo.ensureIndexes()
})

afterAll(async () => {
  await mongo.db.dropDatabase()
  await mongo.disconnect(true)
})

describe(NAME, () => {
  it('should find a shapefile within the zip file', async () => {
    const URL = `http://${NAME}-shp-zip`

    nock(URL)
      .get('/data.zip').reply(200, () => shapefile('data'), {
        'Transfer-Encoding': 'chunked'
      })

    const url = `${URL}/data.zip`
    const {_id} = await upsertLink(url)
    await check.handler({
      data: {
        linkId: _id,
        location: url
      }
    })

    const summary = await getLinkSummary(_id)
    expect(downloadsSnapshot(summary.downloads)).toMatchSnapshot()
  })

  it('should override the downloads if data is different', async () => {
    const URL = `http://${NAME}-shp-zip-caching`

    const url = `${URL}/data.zip`
    const {_id} = await upsertLink(url)

    nock(URL)
      .get('/data.zip')
      .reply(200, () => shapefile('data'), {
        'Transfer-Encoding': 'chunked'
      })

    await check.handler({
      data: {
        linkId: _id,
        location: url
      }
    })

    const firstSummary = await getLinkSummary(_id)
    expect(downloadsSnapshot(firstSummary.downloads)).toMatchSnapshot()

    nock(URL)
      .get('/data.zip')
      .reply(200, () => shapefile('data', 'foo'), {
        'Transfer-Encoding': 'chunked'
      })

    await check.handler({
      data: {
        linkId: _id,
        location: url,
        options: {
          noCache: true
        }
      }
    })

    const lastSummary = await getLinkSummary(_id)
    expect(downloadsSnapshot(lastSummary.downloads)).toMatchSnapshot()
  })
})
