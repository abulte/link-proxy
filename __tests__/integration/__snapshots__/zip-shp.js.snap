// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test-link-proxy-zip-shp should find a shapefile within the zip file 1`] = `
Array [
  Object {
    "archive": true,
    "files": Array [
      Object {
        "digest": "sha384-+1t5TeJcJ5X34OTBKFd9lDXmWfSP4yyKwNCd6HqLGKChrQodibskoUMz8lbBRABG",
        "name": "data.shp",
        "size": 11,
      },
      Object {
        "digest": "sha384-g/OOiDmdzWdrQElweiK218tYBpzU+Sqpg39YDz/OGKo2bHgTMSfs8PTUFsXaKRtU",
        "name": "data.cpg",
        "size": 11,
      },
      Object {
        "digest": "sha384-N0PxnfQbUbtgRHdg85YR5d23TJyCCw9dkMQ2b7sKrMLplKm2yiK2a+bPBWapI+PA",
        "name": "data.dbf",
        "size": 11,
      },
      Object {
        "digest": "sha384-KjIjW8mDGhYle3IvWHrtMZ3Yng0z3375s8hE8w9gcirOr8Ha1mldECb4q04fg4vc",
        "name": "data.prj",
        "size": 11,
      },
      Object {
        "digest": "sha384-FA1fzHfLn30z4w1QOE8IM3GlS45Xvmodk+MQdzWCkx8ceyOzE3aLBV9Gl7XXZJ4e",
        "name": "data.shx",
        "size": 11,
      },
    ],
    "type": "shp",
  },
]
`;
